core/corehttp/readers.go

```diff
@@ -0,0 +1,134 @@
+package corehttp
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/ipfs/go-ipfs/commands/files"
+	"github.com/ipfs/go-ipfs/importer"
+	chunk "github.com/ipfs/go-ipfs/importer/chunk"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	ft "github.com/ipfs/go-ipfs/unixfs"
+)
+
+// helper functions for the gateway handler
+
+func (i *gatewayHandler) newDagFromReader(r io.Reader) (*dag.Node, error) {
+	// TODO(cryptix): change and remove this helper once PR1136 is merged
+	// return ufs.AddFromReader(i.node, r.Body)
+	return importer.BuildDagFromReader(
+		r, i.node.DAG, chunk.DefaultSplitter, importer.BasicPinnerCB(i.node.Pinning.GetManual()))
+}
+
+// add file, or directory recursively
+func (i *gatewayHandler) addFileRecursive(file files.File) (*dag.Node, error) {
+	if file.IsDirectory() {
+		return i.addDirRecursive(file)
+	}
+	return i.newDagFromReader(file)
+}
+
+// add a directory recursively
+func (i *gatewayHandler) addDirRecursive(file files.File) (*dag.Node, error) {
```

Comment:
ok, you're right about progress bars. the commands version tho does do some outputing (there's this outputDag function) but I suppose we can switch over that and not do it from the gateway. The thing about pinning though, is that add from the cli does it by default. I don't think the gateway should replicate that, since the gateway is a hosted service which is presumably being offered by some remote kind soul to benefit whichever local poor soul doesn't have an ipfs daemon running locally. It would seem to be to be bad manners to have that kind remote gateway pin everything that anyone adds, and hence keep it all locally. My understanding of the gateway is that it is meant to be a service offering a conduit to the ipfs network, not a fully featured endpoint. 

as for the hidden files flag, happy to add that. wheres the proper place to put a param like that in a post request? do i restructure the body or is it fine to go in the url?